chia/wallet/wallet_node.py

```diff
@@ -589,41 +602,55 @@ async def receive_state_from_peer(
         if fork_height is not None:
             cache.clear_after_height(fork_height)
 
-        all_tasks = []
+        all_tasks: List[asyncio.Task] = []
+        target_concurrent_tasks: int = 20
+        num_concurrent_tasks: int = 0
 
-        for idx, potential_state in enumerate(items):
+        async def receive_and_validate(inner_state: CoinState, inner_idx: int):
+            nonlocal num_concurrent_tasks
+            assert self.wallet_state_manager is not None
+            assert self.validation_semaphore is not None
+            # if height is not None:
+            async with self.validation_semaphore:
+                try:
+                    if header_hash is not None:
+                        assert height is not None
+                        self.add_state_to_race_cache(header_hash, height, inner_state)
+                        self.log.info(f"Added to race cache: {height}, {inner_state}")
+                    if trusted:
+                        valid = True
+                    else:
+                        valid = await self.validate_received_state_from_peer(inner_state, peer, cache, fork_height)
+                    if valid:
+                        self.log.info(f"new coin state received ({inner_idx + 1} / {len(items)})")
+                        assert self.new_state_lock is not None
+                        async with self.new_state_lock:
+                            await self.wallet_state_manager.new_coin_state([inner_state], peer, fork_height)
+                        if update_finished_height:
+                            await self.wallet_state_manager.blockchain.set_finished_sync_up_to(
+                                last_change_height_cs(inner_state)
+                            )
+                except Exception as e:
+                    tb = traceback.format_exc()
+                    self.log.error(f"Exception while adding state: {e} {tb}")
+                finally:
+                    num_concurrent_tasks -= 1
```

Comment:
This function can terminate without entering the `try` and thus without decreasing the count.  I think a different mechanism would be better, even short of a worker pool (maybe I'll take a pass at that old PR tomorrow).  I'll try to come up with a suggestion.

Reply:
thanks for point that out. However it can only fail if that task is somehow cancelled right? and we never actually cancel that task here.
Feel free to change this one or make a PR for it, i'm sure you can come up with something more robust

Reply:
There are two asserts at the beginning that are outside of the `try:` along with waiting on the semaphore.  I don't know offhand if the semaphore can fail other than cancellation, but the other two can anyways.  I did just get https://github.com/Chia-Network/chia-blockchain/pull/9122 resurrected without much in the way of conflicts.  I'll need to write up tests and then give it some runtime as well.  Assuming we generally like that direction I'm not super worried about getting this entirely generally proper.  Shifting the `try:` up a few lines seems reasonable though.